cirq/circuits/circuit.py

```diff
@@ -120,63 +97,70 @@ def __init__(self,
                 circuit.
             device: Hardware that the circuit should be able to run on.
         """
-        self._moments: List['cirq.Moment'] = []
-        self._device = device
-        self.append(contents, strategy=strategy)
 
     @property
-    def device(self) -> devices.Device:
-        return self._device
-
-    @device.setter
-    def device(self, new_device: 'cirq.Device') -> None:
-        new_device.validate_circuit(self)
-        self._device = new_device
+    @abc.abstractmethod
+    def moments(self) -> Sequence['cirq.Moment']:
+        pass
 
-    def __copy__(self) -> 'Circuit':
-        return self.copy()
+    @property
+    @abc.abstractmethod
+    def device(self) -> devices.Device:
+        pass
 
-    def copy(self) -> 'Circuit':
-        copied_circuit = Circuit(device=self._device)
-        copied_circuit._moments = self._moments[:]
-        return copied_circuit
+    def freeze(self) -> 'cirq.FrozenCircuit':
```

Comment:
`freeze()` and `unfreeze()` do not feel right here. 

I see two problems with having them on `AbstractCircuit`:

1. now the abstract class needs to know about its implementors which is really not typical, and introduces a circular dependency in the type hierarchy - AbstractCircuit is now depending on FrozenCircuit which implements it.
2. Also if ever a new type of circuit would be added, this method (thus the class) would have to change - this is a bit against the [Open Closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) - theoretically you could just plop in a new subclass implement the right methods and be done with it. With the current design you have to somehow "know" that you have to change these two methods as well. 
3. even for discoverability it is confusing: even in the current code you call it a couple of ways `AbstractCircuit.freeze` and `FrozenCircuit.freeze`

My suggestion:  We could make freeze/unfreeze into a protocol: 
`__freeze__`, `__unfreeze__` would be the magic methods: 
```
@abstractmethod
AbstractCircuit.__freeze__
@abstractmethod
AbstractCircuit.__unfreeze__
``` 
And then the two subclasses can implement these magic methods to their liking. 

The freeze function would check for the existence of the private methods like in other Cirq protocols.

Reply:
@balopat, do you imagine other types being freezable/unfreezable besides possible future subclasses of `AbstractCircuit`? Seems like it doesn't merit a full-on protocol because I don't see this as being a very generic operation that we will use over multiple types. Also, FWIW I like instance methods instead of class methods for freezing (e.g. call `my_circuit.freeze()` instead of `AbstractCircuit.freeze(my_circuit)`.

Reply:
Well, technically, I can image a mutable version of Devices, Moments and Operations as well but it is a stretch and I have no evidence for a need to introduce them. So I think you're right about a protocol being overkill here. 

The instance methods make sense - but then would you enforce an implementation for them from the `AbstractCircuit` with an abstract method? I would.

Reply:
Actually I just discovered that `PauliString` has a `frozen()` and `mutable_copy()` method. It might be worthwhile to name them similarly - but maybe two types are already enough to warrant a protocol?

Reply:
+ there is a `freeze_op_tree` helper method that freezes the OP_TREE type.

Reply:
Interesting to have those other examples. I'd still suggest using instance methods here for now, but we could look into making a freeze protocol in the future. I'd want to be sure the semantics are consistent across all the freezable types, e.g. w.r.t. returning new instances vs. changing internal flags to mark things as frozen. The name `mutable_copy` has the advantage over `unfreeze` that it indicates it will make a copy, even if given an object that is already mutable. So perhaps we should use that here instead of `unfreeze`? I don't know which is more important, symmetry between `freeze`/`unfreeze` or indicating to the caller that it makes a copy.